### PR TITLE
Method log_2 optimized to use 7k gas instead of 10.5k gas

### PR DESCRIPTION
**Pull request by k06a**
- Unrolled binary search of top bit
- Avoided branching inside second loop
- Avoided non-uint256 types

### CODE REVIEW

**Comment by k06a** [comment]
Are you going to publish unit tests for your library and configure CI for it?

File: ABDKMath64x64.sol [hunk 1]
```diff
@@ -366,31 +366,25 @@ library ABDKMath64x64 {
...
     uint256 ux = uint256 (x) << 127 - a;
-    for (int128 bit = 0x8000000000000000; bit > 0; bit >>= 1) {
+    for (uint256 bit = 0x8000000000000000; bit > 0; bit >>= 1) {
       ux *= ux;
-      if (ux >=
-        0x8000000000000000000000000000000000000000000000000000000000000000) {
-        ux >>= 128;
-        result += bit;
-      } else ux >>= 127;
+      result |= ((ux >> 255) * bit);
```
Comment (3sGgpQ8H) [inline ABDKMath64x64.sol]: This does not seem optimal: multiplication is used where shift would work.  Also, the calculations are performed even when `ux >> 255` s zero.

Comment (k06a) [reply ABDKMath64x64.sol]: @3sGgpQ8H, but I tested it, branching costs more gas.

**Comment by 3sGgpQ8H** [comment]
Optimizations implemented in slightly different way: https://github.com/abdk-consulting/abdk-libraries-solidity/commit/2b46e39de421ed0a16844f1ec9f2f73d34a55843.  Anyway, thank you so much for your interest and contribution.

**Pull request closed without merging**